src/containers/unicode_point_map.rs

```diff
@@ -0,0 +1,1124 @@
+    pub fn remove(
+        &mut self,
+        c : &char
+    ) -> Option<isize> {
+        let ix = *c as usize;
+
+        if ix < self.vec.len() {
+            let prev = self.vec[ix];
+
+            self.vec[ix] = 0;
+
+            if 0 != prev {
+                self.len -= 1;
+
+                self.total -= prev as i64;
+
+                Some(prev)
+            }
+            else {
+                None
+            }
+        } else {
+            match self.map.remove(c) {
+                Some(v) => {
+
+                    self.len -= 1;
+                    self.total -= v as i64;
+
+                    Some(v)
+                },
+                None => {
+                    None
+                }
+            }
+        }
+    }
+}
+
+// Non-mutating methods
+
+impl UnicodePointMap {
+    /// Returns the number of elements the map can hold without
+    /// reallocation.
+    pub fn capacity(&self) -> usize {
+        self.vec.len() + self.map.capacity()
+    }
+
+    /// Indicates whether a record exists for the given key.
+    pub fn contains_key(
+        &self,
+        c : &char,
+    ) -> bool {
+        let ix = *c as usize;
+
+        if ix < self.vec.len() {
+            self.vec[ix] != 0
+        } else {
+            self.map.contains_key(c)
+        }
+    }
+
+    /// Obtains the count corresponding to the given key, obtaining 0 in the
+    /// case that no such record exists.
+    pub fn get(
+        &self,
+        c : &char,
+    ) -> isize {
+        let ix = *c as usize;
+
+        if ix < self.vec.len() {
+            self.vec[ix]
+        } else {
+            match self.map.get(c) {
+                Some(&v) => v,
+                None => 0,
+            }
+        }
+    }
+
+    /// Indicates whether the instance contains no elements.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.is_empty_()
+    }
+}
+
+/// Iterator for [`UnicodePointMap`].
+pub struct UnicodePointMapIter<'a> {
+    /// Reference to the providing [`UnicodePointMap`] instance.
+    upm : &'a UnicodePointMap,
+    /// (Optional) index into the vector.
+    vec_index : Option<usize>,
+    /// (Optional) interator into the map.
+    map_iter : Option<std::collections::hash_map::Iter<'a, char, isize>>,
+}
+
+impl Iterator for UnicodePointMapIter<'_> {
+    type Item = (char, isize);
+
+    /// Advances the iterator and returns the next value.
+    ///
+    /// Returns `None` when iteration is finished. Further calls continue
+    /// to return `None`.
+    #[inline]
+    fn next(&mut self) -> Option<(char, isize)> {
+        debug_assert!(self.vec_index.is_none() || self.map_iter.is_none());
+
+        if let Some(ix) = &mut self.vec_index {
+            while *ix < self.upm.vec.len() {
+                let count = self.upm.vec[*ix];
+
+                if 0 != count {
+                    let c = *ix;
+                    let c = c as u32;
+                    let c = unsafe { char::from_u32_unchecked(c) };
```

Comment:
I'm not entirely sure this is always safe.  I think all the cases of `c as usize` need to be resolved to ensure `c` was never truncated.
I think I could contrive a case on an embedded platform where this would be violated, but I need to think about it more...